### PR TITLE
Asset Modularization

### PR DESCRIPTION
**Pull request by RobertBColton**
Ok, so the goal here is to modularize the code we use for writing assets to the exe and launching the game. In addition, we also want to write the new code to accept a proto instead of the old backend structures. This pull request creates a new library called `libAssets` to house this code.

### Output/Logging
One thing we should talk about is having output from `libAssets` show in the IDE's output window. Because I didn't want to move any of the output/logging stuff from `CompilerSource` yet, `libAssets` has all of its logging code commented out. This is complicated by the fact that currently RGM is only reading output from `CompileReply` which is populated from `EnigmaCallback`. Since RGM links directly to `libAssets` it would need to hook into its log methods directly somehow.

### Summary of Changes
* Added `plugin_CompileEGMRaw` which functions the same as the old `plugin_CompileEGM` except it doesn't write the asset data for sprites and sounds onto the exe. For that reason, it also doesn't launch the game when it finishes. I did it this way, because changing the signature of `plugin_CompileEGM` would break LateralGM binary compatibility, which we're not ready to do yet.
* Changed the signature of `EnigmaPlugin::BuildGame` to accept the exe and project paths as well as a boolean controlling whether to write the assets. Because the method now accepts the project's file path, the `working_directory` in run and debug mode will be compatible with GM8 again. This is done because emake obviously needs to write the assets and sometimes launch the game, but the GRPC server only needs to build the game and leave these tasks up to the IDE.
* Added an overload for `EnigmaPlugin::BuildGame` for when the exe path is the same as the project path (e.g, empty game because there is no project file).
* Moved the `OptionsParser` instance from emake main to the global scope so that it can be accessed by `EnigmaPlugin::BuildGame` to determine when the `--run` option has been specified.
* Removed emake's double linkage of `libEGM` because I didn't see any reasoning for it and it may also have been bloating the size of it.
* Added linkage of `libAssets` to emake so it can use it to write assets to the game and launch the game.
* Disabled the code in `Proto2ES` that loads, converts, and compresses the data files for sounds, sprites, and backgrounds because it is no longer needed. Because `libAssets` has new code to write the assets, I changed the logic for it to do the loading.
* Created `libAssets` with a makefile and a CMakeLists so it can be built in MSYS2 and by RGM's AppVeyor. It contains a new method `game_write_assets` which is based on Josh's code I extracted from `CompilerSource` but it's slightly different in that it accepts the new proto instead of the old EnigmaStruct. The library also has a `game_launch` method which was extracted from `CompilerSource` and is used by emake and LGM to launch the game consistently.
* Tweaked the `libEGM` makefile to rename `SHARED_SOURCES` to `SHARED_SRC_DIR` which is more consistent with the other makefiles we have that use the shared code. The directory variable is supposed to be `SHARED_SRC_DIR` and `SHARED_SOURCES` is supposed to be a list of the actual sources to compile.
* Moved `bettersystem` to the `shared/` folder so that it can be compiled into `libAssets` so that `game_launch` is able to call `e_execs`.
* Cleaned up a few redundant `string_replace_all` implementations by moving it to `bettersystem` in the shared folder.
* Extracted `compile_write_assets` and `compile_run_game` in `compile.cpp` so I could easily copy it to `libAssets` and so I could make `plugin_CompileEGMRaw` not call it.
* Updated the main makefile to include rules for `libAssets` so it gets built by `make all`. At this point, emake now depends on `libAssets` to write assets to and launch games that are protos.
* Updated the AppVeyor job to include `libAssets.dll` in the blobs zip since emake now depends on it.
* Moved the includes of `windows.h` and `unistd.h` to the top of `bettersystem.cpp` before the `using namespace std;` since the using causes issues when compiling with MSVC and C++17.

### CODE REVIEW

File: CompilerSource/compiler/compile.cpp [hunk 1]
```diff
@@ -148,6 +148,109 @@ inline void write_exe_info(const std::string codegen_directory, const EnigmaStru
   wto.close();
 }
 
+#define irrr() if (res) { idpr("Error occurred; see scrollback for details.",-1); return res; }
```
Comment (JoshDreamland) [inline CompilerSource/compiler/compile.cpp]: Please move these compiler changes to their own PR, if you intend to continue working on those "libassets" files.

File: CompilerSource/compiler/compile.cpp [hunk 2]
```diff
@@ -684,103 +785,13 @@ wto << "namespace enigma_user {\nstring shader_get_name(int i) {\n switch (i) {\
...
-  // Close the game module; we're done adding resources
-  idpr("Closing game module and running if requested.",99);
-  edbg << "Closing game module and running if requested." << flushl;
-  fclose(gameModule);
+  int ret = compile_append_resources(gameFname.c_str(), es);
+  if (ret) return ret;
 
   // Run the game if requested
   if (run_game && (mode == emode_run or mode == emode_debug or mode == emode_design))
   {
```
[the remark this answers is not in the document]
Comment (RobertBColton) [reply CompilerSource/compiler/compile.cpp]: This was resolved at some point, the brace is now on the same line.

**Comment by RobertBColton** [comment]
@JoshDreamland alright, this is ready to go for review!

File: CommandLine/libAssets/assets.cpp [hunk 3]
```diff
@@ -0,0 +1,433 @@
+/** Copyright (C) 2018 Robert B. Colton
```
Comment (JoshDreamland) [inline CommandLine/libAssets/assets.cpp]: 1. You have to include the original license from the code you moved to this file.
2. There's not a compelling reason to duplicate that code, here.
3. You removed all the nice logging mechanisms that code had. The mechanism was relatively well-contained; you could replace it with a no-op logger if you genuinely had nowhere to send that information, but odds are, you can have it stream to your RPC when writing to stdout or LGM isn't good enough.

The `lang_CPP` class was meant to be an implementation of a language export for C++. The other I intended was EGMjs (other repository). If the class depended on global state (such as for logging), that's relatively easy to correct. The answer here is not to duplicate that code and make it this gigantic monolithic file. Please don't do this.

File: shared/bettersystem/bettersystem.h [hunk 4]
```diff
@@ -39,3 +42,16 @@ int e_execsp(std::string cmd, std::string cat1, std::string cat2, std::string pa
 int e_execsp(std::string cmd, std::string cat1, std::string cat2, std::string cat3, std::string path);
 
 //int better_system(std::string,std::string,std::string="",const char* = NULL);
+
+inline std::string string_replace_all(std::string str, const std::string& substr, const std::string& nstr)
```
[the remark this answers is not in the document]
Comment (RobertBColton) [reply shared/bettersystem/bettersystem.h]: Josh, did you miss where I said:
>Cleaned up a few redundant string_replace_all implementations

This is actually 1 less `string_replace_all` I moved it to better system so it's shared. Now what we could do is move it from bettersystem to its own shared string library and delete the redundant `estring.h` version of it too.

That said, yeah I changed it to return a copy of the string instead of mutating. It seems we also happen to have two versions in our source code. Which do you prefer, one that mutates the original string or one that returns a mutated copy of the string?

File: CompilerSource/compiler/compile.cpp [hunk 5]
```diff
@@ -684,104 +782,15 @@ wto << "namespace enigma_user {\nstring shader_get_name(int i) {\n switch (i) {\
...
-
-    if (resourceblock_start < 128) {
-      user << "Compiled game is clearly not a working module; cannot continue" << flushl;
-      idpr("Failed to add resources.",-1); return 13;
+    // Run the game if requested
+    if (run_game && (mode == emode_run or mode == emode_debug or mode == emode_design)) {
+      compile_run_game(gameFname.c_str(), es, mode);
     }
   }
-  else
```
[the remark this answers is not in the document]
Comment (RobertBColton) [reply CompilerSource/compiler/compile.cpp]: I don't know how I feel about this. I mean obviously you were ok with the rc file writing being extracted to helper functions for Win32 and Linux `.desktop` files? I see it as somewhat of an improvement to have the compile method more "modular" and less of a monolithic heap. Regardless, I still have time to fix it, move the code back and minimize the diff, but are you sure that's what you really want?

File: CommandLine/libAssets/assets.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
+/** Copyright (C) 2018 Robert B. Colton
```
[the remark this answers is not in the document]
Comment (RobertBColton) [reply CommandLine/libAssets/assets.h]: Rofl, I don't think that's how copyright works Josh, it's code one way or another, it's not like we're filing a patent on it lol. Also, we have a copyright notice in a lot less copyrightable areas of the source code already. Regardless, that doesn't change the fact I should have included you on the copyright since I am adapting your old code.

File: CommandLine/libAssets/assets.cpp (hunk 3 above)
Comment (RobertBColton) [reply CommandLine/libAssets/assets.cpp]: 1. It is the original license, I was just missing the original author(s) until dbd36a4c7c1e683ee5a691c920c2977a3b938c97
2. We need a version that accepts a proto, not EnigmaStruct, but we're not yet ready to break LGM until the next pull request.
3. I didn't remove them, I commented them out for a second. Regardless, that's a fair point you've made. So you're saying, even when RGM calls the function directly (since it statically links `libAssets`) it should send the log messages back to the server and then to RGM through GRPC? That's going to require some additional restructuring to the logging, because the `BuildGame` RPC is already in the "finished" state when RGM writes the assets, hence there's no more opportunity to read the log messages from CompileReply. That's why I suggested offering a way for RGM to read `libAssets` log messages directly.

>The lang_CPP class was meant to be an implementation of a language export for C++.

Right, I understood this much.

>The answer here is not to duplicate that code and make it this gigantic monolithic file.

Are you talking about just the logging or the logging and asset writing? You see, you should understand by now the idea here is to let RGM write the assets to the game. If that code is embedded into `lang_cpp` then `lang_cpp` and the rest of `CompilerSource` also has to be linked into RGM for it to be able to write the assets.

What else do you propose? We've already discussed your idea of distributing an emake binary with RGM but we preferred the single-executable approach. I need additional feedback from you before I can continue.
